### DISCUSSION TITLE
Device could not found error for *device command

### DISCUSSION THREAD

**Discussion by emsumithn** [mtconnect/cppagent#411] [Q&A] [answered]
I am currently implementing agent commands as a part of adapter stream.

I want to rewrite uuid or name in device file by using command * device: <uuid|name> which will be sent from adapter to agent.
Believe that * device command will replace default value defined in device file during runtime. 
VersionDeviceXml is set to true in agent config file to support this.
"
Not sure why am i getting this error that "Cannot find device for name device1"  Is there any other configuration to be done for this? 

Any guidance could be of great help
Thank you

**Comment by wsobel** [comment, marked as answer]
The `* device: ` sets the device associated with the adapter to the device identified by the `uuid` or `name`

The command, `* uuid: ` will do what you want. I checked and it is not documented. This is our fault and I will fix it in the README.md. I'll add it.

If you turn on versioning, the Device.xml fill should update to with the new uuid once it is received. 

Sorry for the oversight.

- Will

**Reply by wsobel** [reply]
Created issue #412 to address the documentation oversight.

**Reply by emsumithn** [reply]
@wsobel Thank you for the reply and updates.

I tried to change UUID by sending command `* uuid: 9876544`.
`VersionDeviceXml` is set to true .
`EnableSourceDeviceModels` is tried with both true and false scenario.

In the logs, i can see that command is being processed

```
 2024-02-15T04:43:36.439322Z (0x00007f807ede9b38) [trace] Connector::reader->Connector::parseSocketBuffer->Connector::processLine: (10.0.0.99:9800) Received line: * uuid: 9876544
 
 2024-02-15T04:43:36.439347Z (0x00007f807ede9b38) [debug] Connector::reader->Connector::parseSocketBuffer->Connector::processLine->ShdrAdapter::processData->ShdrAdapter::protocolCommand: Processing command: uuid: 9876544
```

But i don't see any changes in REST API output, it shows default uuid defined in the device file. What could be the potential issue here?


> The * device:  sets the device associated with the adapter to the device identified by the uuid or name

Referring to this, so there is no way to rewrite device name with some random string values , but we can set default device for an adapter among the defined devices in configuration file using `* device` command. Is my understanding right ?

**Reply by wsobel** [reply]
2 questions:

- Can you upload your agent config file?
- What is the use case or issue you're trying to solve?

I'd like to better understand what's happening and what you want to achieve to better address the solution. 

I'll test the agent and make sure the uuid command is working properly. Otherwise, it could be another setting in the agent config that could be conflicting. 

At present you can't change the device name, only the uuid. I think this is historical and nobody has asked to change the machine name. The uuid can come from the machine because they used a generic device xml and needed to set it to the serial number plus the mfg and model. Something like that. 

I'm interested in your use case because this may be of general interest. 

Thanks
- W

**Reply by emsumithn** [reply]
> At present you can't change the device name, only the uuid

Thank you for the clarification.

`* uuid` command is still not working, here is Config file for your reference:
```
Devices = device_saw.xml
ServerIP = 0.0.0.0
Port = 5000
AllowPut = false
AllowPutFrom = localhost

ReconnectInterval = 1000
WorkerThreads = 2
BufferSize = 17
MaxAssets = 8096

SchemaVersion = 2.0
ShdrVersion = 2.0 
JsonVerson = 2.0

SuppressIPAddress = true
IgnoreTimestamps = true
MonitorConfigFiles = true
ConversionRequired = false
UpcaseDataItemValue = true
EnableSourceDeviceModels = false
VersionDeviceXml = true
Pretty = true
TlsOnly = false


Adapters {
    Saw {
        Protocol = shdr
        RealTime = false
        Host = 10.0.0.4
        Port = 9800
    }
}
```

Thank you for your quick response.

**Reply by wsobel** [reply]
Thanks for the info. This will make it easier to reproduce. 

I'm still curious about your use case. 

Best
W

**Reply by emsumithn** [reply]
Able to change UUID with `* uuid` command. 
There was a configuration `PreserveUUID`  , by default this will be true.

`PreserveUUID = false`
Adding the above to config solves the issue without ` VersionDeviceXml ` also.

Thank you
